### PR TITLE
`Exam mode`: Fix an issue when submitting complaints

### PR DESCRIPTION
**Pull request by Strohgelaender**
#### General
- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context
#6875 fixed a special case where complaints were not possible but missed that check on the server side.

### Description
Fix the server side of this check

### Steps for Testing
Prerequisites:
- 1 Instructor
- 2 Student
- 1 Exam

1. Edit the course to have some low complaint text limit (e.g 10 characters)
2. Submit a exam complaint which uses more than 10 characters
3. Everything should work
4. Course complains should still be limited to 10 characters.
5. Also test exam complaints with course complaints disabled.

#### Performance Review
- [x] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [ComplaintResponseService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5222/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/ComplaintResponseService.html) | 90% | ✅ |
| [ComplaintService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5222/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/ComplaintService.html) | 85% | ✅ |

### CODE REVIEW

**Review by JohannesStoehr — DISMISSED** [review]
Code looks good to me

File: src/test/java/de/tum/in/www1/artemis/assessment/AssessmentComplaintIntegrationTest.java [hunk 1]
```diff
@@ -990,4 +1014,60 @@ void submitComplaintForExerciseComplaintNotExceededTextLimit() throws Exception
...
+    @Test
+    @WithMockUser(username = TEST_PREFIX + "student1", roles = "USER")
+    void submitComplaintForExam_courseComplaintsEnabled_exceededCourseLimit_success() throws Exception {
+        TextExercise examExercise = textExerciseUtilService.addCourseExamExerciseGroupWithOneTextExercise();
+        Course examCourse = examExercise.getCourseViaExerciseGroupOrCourseMember();
+        examCourse = courseUtilService.updateCourseComplaintTextLimit(examCourse, 25);
+        // enable course complaints
+        examCourse.setMaxComplaintTimeDays(3);
+        courseRepository.save(examCourse);
+        // 26 characters, exceeds course limit but lower than 2000 --> allowed
```
Comment (pal03377) [inline src/test/java/de/tum/in/www1/artemis/assessment/AssessmentComplaintIntegrationTest.java:1027]:
```suggestion
        // 26 characters, exceeds course limit but lower than 2000 --> allowed for exam exercise
```

File: src/test/java/de/tum/in/www1/artemis/assessment/AssessmentComplaintIntegrationTest.java [hunk 2]
```diff
@@ -990,4 +1014,60 @@ void submitComplaintForExerciseComplaintNotExceededTextLimit() throws Exception
...
+
+    @Test
+    @WithMockUser(username = TEST_PREFIX + "student1", roles = "USER")
+    void submitComplaintForExam_courseComplaintsDisabled_exceededTextLimit() throws Exception {
+        TextExercise examExercise = textExerciseUtilService.addCourseExamExerciseGroupWithOneTextExercise();
+        Course examCourse = examExercise.getCourseViaExerciseGroupOrCourseMember();
+        // disable course complaints
+        examCourse.setMaxComplaintTimeDays(0);
+        courseRepository.save(examCourse);
+        // 2004 characters
```
Comment (pal03377) [inline src/test/java/de/tum/in/www1/artemis/assessment/AssessmentComplaintIntegrationTest.java:1057]:
```suggestion
        // 2004 characters (4 over the limit of 2000)
```

File: src/main/java/de/tum/in/www1/artemis/service/ComplaintResponseService.java [hunk 3]
```diff
@@ -194,7 +194,11 @@ public ComplaintResponse resolveComplaint(ComplaintResponse updatedComplaintResp
             final Course course = originalComplaint.getResult().getParticipation().getExercise().getCourseViaExerciseGroupOrCourseMember();
 
             // Check whether the complaint text limit is exceeded
-            if (course.getMaxComplaintResponseTextLimit() < updatedComplaintResponse.getResponseText().length()) {
+            int maxLength = course.getMaxComplaintResponseTextLimit();
+            if (originalComplaint.getResult().getParticipation().getExercise().isExamExercise()) {
+                maxLength = Math.max(maxLength, Course.DEFAULT_COMPLAINT_TEXT_LIMIT);
+            }
```
[the remark this answers is not in the document]
Comment (pal03377) [reply src/main/java/de/tum/in/www1/artemis/service/ComplaintResponseService.java:200]: I would also be fine with how it currently implemented, but personally I would tend towards making the limit part of the exercise. What do you think?

**Review by pal03377 — CHANGES_REQUESTED** [review]
I have a suggestion for you to consider about the architecture of this change as well as some minor comment improvement suggestions. 

Otherwise, everything looks good & especially the tests look really clean! 🎉

**Review by pal03377 — DISMISSED** [review]
Code looks good & the small refactoring since my last review is quite sensible 🎉

**Review by terlan98 — DISMISSED** [review]
Tested on TS5 with low complaint text limit and disabled complaints. Works as expected 👍 

Found a small, non-blocking typo in the code when going through Paul's suggestions. I did not do a full code review, though.

**Review by terlan98 — APPROVED** [review]

**Review by pal03377 — APPROVED** [review]

**Review by MaximilianAnzinger — APPROVED** [review]
LGTM

**Pull request merged**